api/handlers.go

```diff
@@ -0,0 +1,156 @@
+}
+
+// the health check endpoint simply returns a 200
+func HealthCheck(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(200) //nolint:errcheck
+}
+
+// GetCertificateRequests returns all of the Certificate Requests
+func GetCertificateRequests(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		certs, err := env.DB.RetrieveAll()
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		body, err := json.Marshal(certs)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		if _, err := w.Write(body); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// PostCertificateRequest creates a new Certificate Request, and returns the id of the created row
+func PostCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		csr, err := io.ReadAll(r.Body)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		id, err := env.DB.Create(string(csr))
+		if err != nil {
+			if strings.Contains(err.Error(), "UNIQUE constraint failed") {
+				logError("given csr already recorded", 400, w)
+				return
+			} else {
+				logError(err.Error(), 400, w)
+				return
+			}
+		}
+		w.WriteHeader(201)
+		if _, err := w.Write([]byte(strconv.FormatInt(id, 10))); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// GetCertificateRequests receives an id as a path parameter, and
+// returns the corresponding Certificate Request
+func GetCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		id := r.PathValue("id")
+		cert, err := env.DB.Retrieve(id)
+		if err != nil {
+			if err.Error() == "sql: no rows in result set" {
+				logError(err.Error(), 400, w)
+				return
+			}
+			logError(err.Error(), 500, w)
+			return
+		}
+		body, err := json.Marshal(cert)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		if _, err := w.Write(body); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// DeleteCertificateRequest handler receives an id as a path parameter,
+// deletes the corresponding Certificate Request, and returns a 204 on success
+func DeleteCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		id := r.PathValue("id")
+		err := env.DB.Delete(id)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		w.WriteHeader(204)
+	}
+}
+
+// PostCertificate handler receives an id as a path parameter,
+// and attempts to add a given certificate to the corresponding certificate request
+func PostCertificate(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		cert, err := io.ReadAll(r.Body)
+		if err != nil {
+			logError(err.Error(), 400, w)
+			return
+		}
+		id := r.PathValue("id")
+		insertId, err := env.DB.Update(id, string(cert))
+		if err != nil {
+			logError(err.Error(), 400, w)
```

Comment:
Should we return 404 here instead?

Reply:
I'd prefer not to have explicit 404's in general, go will return a 404 for paths that don't exist. In this case, if there's an error, it will be either because of the validation of the cert or a connection error. I've differentiated between the errors in the certdb side so this part reads a bit cleaner.